src/redux/actions/authActions.js

```diff
@@ -0,0 +1,143 @@
+import fetch from 'cross-fetch';
+
+export const SIGNUP = 'SIGN_UP';
+export const SIGNUP_REQUEST = 'REQUEST_SIGNUP';
+export const SIGNUP_FAILURE = 'SIGNUP_FAILURE';
+export const SIGNUP_SUCCESS = 'SIGNUP_SUCCESS';
+export const LOGOUT = 'LOGOUT';
+export const LOGIN = 'LOGIN';
+export const LOGIN_REQUEST = 'REQUEST_LOGIN';
+export const LOGIN_FAILURE = 'LOGIN_FAILURE';
+export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';
+export const LOGOUT_REQUEST = 'LOGOUT_REQUEST';
+export const LOGOUT_FAILURE = 'LOGOUT_FAILURE';
+export const LOGOUT_SUCCESS = 'LOGOUT_SUCCESS';
+
+function requestSignup(user) {
+  return {
+    type: SIGNUP_REQUEST,
+    user: user,
+    receivedAt: Date.now()
+  };
+}
+
+function signupSuccess(user, json) {
+  return {
+    type: SIGNUP_SUCCESS,
+    user: json,
+    receivedAt: Date.now()
+  };
+}
+
+function signupFailure() {
+  return {
+    type: SIGNUP_FAILURE,
+    user: null,
+    receivedAt: Date.now()
+  };
+}
+
+function requestLogin(user) {
+  return {
+    type: LOGIN_REQUEST,
+    user: user,
+    receivedAt: Date.now()
+  };
+}
+
+function loginSuccess(user, json) {
+  return {
+    type: LOGIN_SUCCESS,
+    user: json,
+    receivedAt: Date.now()
+  };
+}
+
+function loginFailure() {
+  return {
+    type: LOGIN_FAILURE,
+    user: null,
+    receivedAt: Date.now()
+  };
+}
+
+function requestLogout() {
+  return {
+    type: LOGOUT_REQUEST,
+    user: null,
+    receivedAt: Date.now()
+  };
+}
+
+function logoutSuccess() {
+  return {
+    type: LOGOUT_SUCCESS,
+    user: null,
+    receivedAt: Date.now()
+  };
+}
+
+function logoutFailure() {
+  return {
+    type: LOGOUT_FAILURE,
+    user: null,
+    receivedAt: Date.now()
+  };
+}
+
+export function signup(user) {
+  return dispatch => {
+    dispatch(requestSignup(user)); // Signup request process has begun...
+    return fetch('http://localhost:4201/signup', {
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      method: 'post',
+      body: JSON.stringify(user)
+    })
+      .then(response => response.json())
+      .then(json => {
+        if (json.success === false) dispatch(signupFailure());
+        else {
+          localStorage.setItem('token', json.token);
+          dispatch(signupSuccess(user, json));
+        }
+      });
+  };
+}
+
+export function login(user) {
+  return dispatch => {
```

Comment:
If I'm understanding this correctly, we call login and it returns the resulting json response (if there is one—we should handle when there is no response at all in the error handling). But shouldn't all effects be caused by the dispatched actions from this method? That is, the action requestLogin, loginFailure, and loginSuccess are the only things that should cause an effect after calling this method, so this method might not need to return anything.

Also, because this method should not really have an "effect" on anything else, including the redux store (though it does this indirectly by dispatching things), I am very much in favour of pulling this method into another file—maybe even outside of the actions folder.

Reply:
Your understanding is correct. That is how I understood NgRx's data flow, but the official Redux tutorial confused me a little bit... There is still an emphasis on that API calls should never be made within reducers, but the official tutorials place these methods inside Action files - see [Redux Async Actions](https://redux.js.org/advanced/async-action) to get an idea of what I'm talking about. The lines between Actions, Reducers, and Effects is not as clear to me as it was in NgRx (being that there doesn't seem to be Effects files in Redux?). I'll spend this afternoon looking at it again with fresh eyes and see what I can learn and what can be improved upon here.

Reply:
Here's a stackoverflow post that touches on this as well [How to properly make REST calls from ReactJS + Redux application?](https://stackoverflow.com/questions/38728884/how-to-properly-make-rest-calls-from-reactjs-redux-application)

Reply:
Here's another article that elaborates on actions performing asynchronous actions [Where and When to Fetch Data With Redux](https://daveceddia.com/where-fetch-data-redux/)

Reply:
Wacky! That's cool. I think what you've done makes sense, in that case!